Lib/glyphsLib/builder/glyph.py

```diff
@@ -63,6 +68,47 @@ def to_ufo_glyph(self, ufo_glyph, layer, glyph):  # noqa: C901
                 layers = []
                 for layerId, colorId in layerMapping:
                     layers.append((glyph.layers[layerId], colorId))
+                self._color_palette_layers.append(((glyph, layer), layers))
+
+        if self.minimal:
+            # The other kind of color layers supports solid colors and
+            # gradients among other things, and we use it to build COLRv1
+            # table.
+            # For each color layer, we collect paths that has the same
+            # attributes. If there is only one group of such paths in a layer,
+            # the layer is used without changes, otherwise we make a clone of
+            # the layer for each group with only the paths in this group. We do
+            # this splitting because a COLRv1 layer can’t have multiple
+            # gradients or colors.
+            color_layers = [
+                l
+                for l in glyph.layers
+                if l.attributes.get("color")
+                and l.associatedMasterId == layer.associatedMasterId
+            ]
+            if color_layers:
+                import json
+
+                layers = []
+                for color_layer in color_layers:
+                    # XXX: What to do with components? Glyphs does not allow
+                    # setting attributes on them.
+                    assert not color_layer.components
+
+                    groups = {}
+                    for path in color_layer.paths:
+                        # XXX we need a smarter way to group paths with
+                        # identical attributes.
+                        key = json.dumps(path.attributes)
```

Comment:
using json to get a string representation of the object is clever, but if it's just to use attributes as key of the groups, you could turn them in tuples of tuples I guess.

Reply:
The problem is that attributes can contain lists and other tables.

Reply:
What I really want is to group paths with same attributes together. Using dictionary is probably a hack in the first place, but I don’t know a better way.

Reply:
have you tried itertools.groupby?

Reply:
I didn’t know about it, thanks. Taking a look, I’m not sure it can help as it requires the list to be already sorted, and I wouldn’t know how to sort the attributes.

Reply:
I just keep the previous attributes in a variable and compare the current. You should not collate all layers with the same attributes as there might be something between them.

Reply:
Interesting! This would allow me to use itertools.groupby then.